wnfs/src/private/file.rs

```diff
@@ -813,6 +655,264 @@ impl PrivateFileContent {
+                .put_encrypted(&name, Some(content_cid), store)
+                .await?;
+        }
+
+        Ok(PrivateForestContent {
+            key,
+            base_name: forest.get_accumulated_name(&base_name),
+            block_count,
+            block_content_size: MAX_BLOCK_CONTENT_SIZE,
+        })
+    }
+
+    /// Like `new`, but allows streaming in the content.
+    ///
+    /// See `new` for more information.
+    pub async fn new_streaming(
+        file_name: &Name,
+        mut content: impl AsyncRead + Unpin,
+        forest: &mut impl PrivateForest,
+        store: &impl BlockStore,
+        rng: &mut impl CryptoRngCore,
+    ) -> Result<Self> {
+        let (key, base_name) = Self::prepare_key_and_base_name(file_name, rng);
+
+        let mut block_index = 0;
+
+        loop {
+            let mut current_block = vec![0u8; MAX_BLOCK_SIZE];
+            let nonce = SnapshotKey::generate_nonce(rng);
+            current_block[..NONCE_SIZE].copy_from_slice(nonce.as_ref());
+
+            // read up to MAX_BLOCK_CONTENT_SIZE content
+
+            let content_end = NONCE_SIZE + MAX_BLOCK_CONTENT_SIZE;
+            let (bytes_written, done) =
+                utils::read_fully(&mut content, &mut current_block[NONCE_SIZE..content_end])
+                    .await?;
+
+            // truncate the vector to its actual length.
+            current_block.truncate(bytes_written + NONCE_SIZE);
+
+            let tag = key.encrypt_in_place(&nonce, &mut current_block[NONCE_SIZE..])?;
+            current_block.extend_from_slice(tag.as_ref());
+
+            let content_cid = store.put_block(current_block, CODEC_RAW).await?;
+
+            let name = Self::create_block_name(&key, block_index, &base_name);
+            forest
+                .put_encrypted(&name, Some(content_cid), store)
+                .await?;
+
+            block_index += 1;
+
+            if done {
+                break;
+            }
+        }
+
+        Ok(PrivateForestContent {
+            key,
+            base_name: forest.get_accumulated_name(&base_name),
+            block_count: block_index,
+            block_content_size: MAX_BLOCK_CONTENT_SIZE,
+        })
+    }
+
+    /// Load some previously stored keys & pointers to encrypted private forest content
+    /// from given metadata key.
+    pub fn load_from_metadata(metadata: &Metadata, key: &str) -> Result<Self> {
+        let wrapped: MetadataContentCapsule<Self> = metadata
+            .get_deserializable(key)
+            .ok_or_else(|| FsError::NotFound)??;
+
+        Ok(match wrapped {
+            MetadataContentCapsule::PrivateForestContent(content) => content,
+        })
+    }
+
+    /// Store these pointers & key to some private forest content in given metadata under given key.
+    pub fn store_in_metadata(&self, metadata: &mut Metadata, key: &str) -> Result<()> {
+        metadata.put_serializable(key, MetadataContentCapsule::PrivateForestContent(&self))?;
+        Ok(())
+    }
```

Comment:
I'm thinking this function really belongs to PrivateFile. So we wouldn't need separate fns like `get_metadata_mut_rc` that if missed can lead to catastrophic failures.
A function like `PrivateFile::upsert_metadata`, perhaps an Entry API. The point is to let file handle updating the metadata since it is the custodian and it best knows what it needs to be in the correct state.